resource/key.go

```diff
@@ -0,0 +1,133 @@
+package resource
+
+import (
+	"strings"
+
+	"github.com/giantswarm/microerror"
+	"k8s.io/apimachinery/pkg/api/meta"
+)
+
+const (
+	// labelObject is a label added to all objects created by resoucres of
+	// this package. It is used for deletion purposes. The value of the
+	// label is namespace and name of watched object in format
+	// "${namespace}-${name}". Together with labelResource it allows to
+	// find objects created by a resource.
+	labelObject = "operatorkit.giantswarm.io/object"
+	// labelResource is label added to all objects created by resources of
+	// this package. It is used for deletion purposes. The value of the
+	// label is the Name property of the setting resource. Together with
+	// labelObject it allows to find objects created by a resource.
+	labelResource = "operatorkit.giantswarm.io/resource"
+)
+
+type assertLabelsType interface {
+	printNameType
+	GetLabels() map[string]string
+}
+
+func assertLabels(obj assertLabelsType, watchedObj interface{}, resourceName string) error {
+	desired, err := newLabels(watchedObj, resourceName)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	for k, v := range desired {
+		current := obj.GetLabels()
+		if current == nil {
+			return microerror.Maskf(executionFailedError, " %#q object %#q label is not set", newPrintName(obj), k)
+		}
+		value, ok := current[k]
+		if !ok {
+			return microerror.Maskf(executionFailedError, " %#q object %#q label is not set", newPrintName(obj), k)
+		}
+		if value != v {
+			return microerror.Maskf(executionFailedError, " %#q object %#q label has value %#q but want %#q", newPrintName(obj), k, value, v)
+		}
+	}
+
+	return nil
+}
+
+func newLabels(watchedObj interface{}, resourceName string) (map[string]string, error) {
+	var labelObjectValue string
+	{
+		accessor, err := meta.Accessor(watchedObj)
```

Comment:
IMO it would be good to do these conversions as early as possible and then being able to pass more specific arguments to the helper functions down the road. This would lead to leaner functions and more specific interfaces and it would be clearer what we are actually doing.